### ISSUE TITLE
Bug in translations: fix inheritance template filename in case of includes from different cross sources between theme, modules, ...

### ISSUE THREAD

**Issue by anaohayo** [PrestaShop/PrestaShop#32770]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

In the smartyfront.config.inc.php file there is a comment to this effect:
line 177: 
`// fix inheritance template filename in case of includes from different cross sources between theme, modules, ...`

### Expected behavior

The solution can be replace these lines 177 - 181:
```
$filename = $smarty->template_resource;
if (!isset($smarty->inheritance->sourceStack[0]) || $filename === $smarty->inheritance->sourceStack[0]->resource) {
	$filename = $smarty->source->name;
}
```
Reblaced by:

```
$filename = $smarty->template_resource;
if ($params['mod'] || !isset($smarty->inheritance->sourceStack[0]) || $filename === $smarty->inheritance->sourceStack[0]->resource) {
	$filename = $smarty->source->name;
}
```

### Steps to reproduce

1. Create a new module and a FrontController.
2. In tpl use the old system of translations. example {l s='Hola' mod='name_of_my_module'}
3. Create translations files ej: translations en.php
4. Run web in english language
5. You can view, that the text is in spanish.

### PrestaShop version(s) where the bug happened

1.7.8.9

### PHP version(s) where the bug happened

7.4

### Your company or customer's name goes here (if applicable).

Ohayo

**Comment by AchrafKchaou1991** [comment]
Hello @anaohayo ,

Thanks for reporting this issue!
Ping @PrestaShop/prestashop-core-developers : Could someone please try to reproduce the issue, it's too technical!

Thanks in advance!

**Comment by dnkhack** [comment]
@anaohayo @AchrafKchaou1991 

I tested. All work fine without modification smartyfront.config.inc.php.

Made the module with the hook in displayBanner. 
All environment is Spain, country, browser, IP, and website with multilanguage (thanks @Luifer-dnk). I attached a module for researching @anaohayo 

[issue32770.zip](https://github.com/PrestaShop/PrestaShop/files/11657052/issue32770.zip)

![vlcsnap-2023-06-05-22h42m03s942](https://github.com/PrestaShop/PrestaShop/assets/35419462/01e55f2e-f92a-48ba-9bd3-67f65611baa1)

![image](https://github.com/PrestaShop/PrestaShop/assets/35419462/2621791f-dd74-41de-9745-6754ca4a8695)

![image](https://github.com/PrestaShop/PrestaShop/assets/35419462/63cce4b8-4b4e-47e6-8aab-d390cd074d3c)

**Comment by anaohayo** [comment]
You have to try with a tpl that is executed not through the php of the module but through a FrontController of the module. I send you screenshots of what happens to me without the modification of smartfront.config.inc.php.

FrontController:
![image](https://github.com/PrestaShop/PrestaShop/assets/117908103/fd80b6f6-769b-4402-9b0b-8af0ccda6902)

Tpl:
![image](https://github.com/PrestaShop/PrestaShop/assets/117908103/abf7fe70-4996-49a2-81a5-9aeb5f042b69)

Translation:
![image](https://github.com/PrestaShop/PrestaShop/assets/117908103/ae42e9d0-8a10-485a-a809-bcd0c1e405ed)

Front in english WITHOUT the modification of smartfront.config.inc.php:
![image](https://github.com/PrestaShop/PrestaShop/assets/117908103/bc4d52c5-02ae-4631-a7d4-9856cc2ef351)

Front in english WITH the modification of smartfront.config.inc.php:
![image](https://github.com/PrestaShop/PrestaShop/assets/117908103/8fa67949-426e-402e-ae5b-3b0367fe1c8f)

**Comment by dnkhack** [comment]
@anaohayo This is not the issue. By default, you have to use **English** as **ID** for the Translator class. And translate for your language!

**Comment by anaohayo** [comment]
I have put the case of English, but what if I don't want to translate in English? if I only want the Italian and Spanish languages I can't translate? I don't see it logical because you give the option to delete the English language from prestashop.

**Comment by anaohayo** [comment]
On the other hand, the documentation states that this is a recommendation, not an obligation (screenshots). As you can see the solution is simple and not aggressive, I don't understand why there is so much criticism, especially when you have a comment pointing out that you have bugs.

![image](https://github.com/PrestaShop/PrestaShop/assets/117908103/4a6e7fec-b442-4cf0-91dd-04533d4e060f)

**Comment by dnkhack** [comment]
@anaohayo Yes, on your provided screenshot you see "will be shown in its original language. Because of this, we..."

`By default, you have to use English as ID` "and then translating..."

And forget Old Translation System. The entity of New is to save developers and modules customers time.

Have a good coding. 😉

**Comment by dnkhack** [comment]
@anaohayo I'm not a Prestashop core developer. I'm simply a coder like you. Came here to provide my bugs and issue. Don't be angry. Prestashop very cool platform with reusable components based on rules and doctrines. PHP always rules, coding standards, and PSR. 
So just accept it, all right?

**Comment by anaohayo** [comment]
@dnkhack Now I understand, I took you for a prestashop developer. I will wait for the response from the developers. Prestashop is the tool I work with every day, and so far it is open source so I can collaborate (something I have already done with other profiles). With this issue I am trying to improve the tool, also the new translation system is still not very good at creating translated multi-language modules, that's why the old system is still being used. I respect and value your collaboration, but in this case I can see that you lack knowledge on the subject. On my site I have already fixed it and all the translations work correctly, I leave it to the experts whether they add the correction or not.

**Comment by hibatallahAouadni** [comment]
Hello @anaohayo 

Do you still encounter the issue with 8.1.x? Please check and feedback.

Thanks!

**Comment by anaohayo** [comment]
@hibatallahAouadni The error occurs in version 1.7.X.X.X, in version 8 I don't know because my shop is still in that subversion.

**Comment by boo-code** [comment]
Not reproducible on 9.2.0 (hummingbird), across six constructions of the reported setup.

The code is unchanged since 1.7: `config/smartyfront.config.inc.php:207-211`, identical in 1.7.8.11, 8.0.0, 9.0.0 and 9.2.x.

I built a module with a `ModuleFrontController` calling `setTemplate('module:<mod>/views/templates/front/<tpl>')`, a template using `{l s='Los secretos de la playa' mod='<mod>'}`, and a `translations/en.php` keyed exactly as `AdminTranslationsController` writes it. Measured `$filename` / `$basename` at the guard with the FO rendering:

| construction | template_resource | source->name | sourceStack[0] | basename | rendered |
| --- | --- | --- | --- | --- | --- |
| plain controller tpl | `module:<mod>/.../display.tpl` | `<mod>/.../display.tpl` | (none) | `display` | translated |
| tpl `{extends 'page.tpl'}` | `layouts/layout-both-columns.tpl` | `<mod>/.../display.tpl` | equal to resource | `display` | translated |
| `{include}` partial inside `{extends}` | `module:<mod>/.../partial.tpl` | `<mod>/.../partial.tpl` | differs | `partial` | translated |
| 3-level `display` -> `middle` -> `page` | `layouts/layout-both-columns.tpl` | `<mod>/.../display.tpl` | equal to resource | `display` | translated |
| string living in the INTERMEDIATE `middle.tpl` | `layouts/layout-both-columns.tpl` | `<mod>/.../middle.tpl` | equal to resource | `middle` | translated |
| template named `my-controller.tpl` | `module:<mod>/.../my-controller.tpl` | `<mod>/.../my-controller.tpl` | (none) | `my-controller` | translated |

The guard resolves to the file that CONTAINS the string in all six, which is what the extractor keys by (`AdminTranslationsController:811`, `strtolower(substr(basename($file), 0, -4))`). That holds structurally in both branches:

- inheritance: `template_resource` is the outermost layout and equals `sourceStack[0]`, so the guard fires and takes `source->name`, which Smarty tracks per compiled unit and is therefore the containing file, not the entry template. Row 5 is the direct test of that: the string sits in `middle.tpl` while the entry was `display.tpl`, and the basename came out `middle`.
- `{include}`: the guard does not fire, but `template_resource` IS the included file. It differs from `source->name` only by the `module:` prefix, which `basename()` strips.

So in these six shapes the proposed `$params['mod'] ||` changes nothing, because `source->name` already equals what the condition would select.

Two notes on the thread. The 2023-06-05 test was a `displayBanner` hook, not the `ModuleFrontController` in the report, so it did not exercise the reported path - that is why the two of you did not converge. And the "use English as ID" point is orthogonal: the key is `md5($source)` regardless of language, so a Spanish source string is not itself the failure.

The remaining variable I could not exclude is the theme prefix. The extractor writes `<{mod}<theme>>` when a theme is selected and `<{mod}prestashop>` otherwise (`AdminTranslationsController:816-820`), and the lookup tries `<{mod}` . `_THEME_NAME_` . `>` then `<{mod}prestashop>` (`classes/Translate.php:121-122`). A `translations/en.php` holding only keys for one theme misses both when the shop runs another, and falls back to the source string - which is exactly the reported symptom, and matches "cross sources between theme, modules" better than the basename does. My fixture used the `prestashop` key, which always matches.

@anaohayo if you still have the shop: which theme was active, and do the keys in the module's `translations/en.php` read `<{my_module}prestashop>my-controller_...` or `<{my_module}<sometheme>>my-controller_...`? If it is the second and the active theme differs, that is the failure, and the fix is not in this guard.
